pennylane/transforms/qcut/montecarlo.py

```diff
@@ -442,102 +461,84 @@ def circuit(x):
+    """
+    A subclass with a custom __call__ method. The custom QNode transform returns an instance
+    of this class.
+    """
 
+    def __call__(self, *args, **kwargs):
         shots = kwargs.pop("shots", False)
-        shots = shots or qnode.device.shots
+        shots = shots or self.device.shots
 
         if shots is None:
             raise ValueError(
                 "A shots value must be provided in the device "
                 "or when calling the QNode to be cut"
             )
 
-        qnode.construct(args, kwargs)
-        tapes, processing_fn = self.construct(qnode.qtape, *targs, **tkwargs, shots=shots)
+        # find the qcut transform inside the transform program and set the shots argument
+        qcut_tc = [
+            tc for tc in self.transform_program if tc.transform.__name__ == "cut_circuit_mc"
+        ][-1]
+        qcut_tc._kwargs["shots"] = shots
 
-        interface = qnode.interface
-        execute_kwargs = getattr(qnode, "execute_kwargs", {}).copy()
-        max_diff = execute_kwargs.pop("max_diff", 2)
-        max_diff = transform_max_diff or max_diff
+        kwargs["shots"] = 1
+        return super().__call__(*args, **kwargs)
 
-        gradient_fn = getattr(qnode, "gradient_fn", qnode.diff_method)
-        gradient_kwargs = getattr(qnode, "gradient_kwargs", {})
 
-        if interface is None or not self.differentiable:
-            gradient_fn = None
-
-        execute_kwargs["cache"] = False
+@cut_circuit_mc.custom_qnode_transform
+def _qnode_transform_mc(self, qnode, targs, tkwargs):
+    """Here, we overwrite the QNode execution wrapper in order
+    to access the device wires."""
+    if tkwargs.get("shots", False):
+        raise ValueError(
+            "Cannot provide a 'shots' value directly to the cut_circuit_mc "
+            "decorator when transforming a QNode. Please provide the number of shots in "
+            "the device or when calling the QNode."
+        )
 
-        res = qml.execute(
-            tapes,
-            device=qnode.device,
-            gradient_fn=gradient_fn,
-            interface=interface,
-            max_diff=max_diff,
-            override_shots=1,
-            gradient_kwargs=gradient_kwargs,
-            **execute_kwargs,
+    if "shots" in inspect.signature(qnode.func).parameters:
+        raise ValueError(
+            "Detected 'shots' as an argument of the quantum function to transform. "
+            "The 'shots' argument name is reserved for overriding the number of shots "
+            "taken by the device."
         )
 
-        out = processing_fn(res)
-        if isinstance(out, list) and len(out) == 1:
-            return out[0]
-        return out
+    tkwargs.setdefault("device_wires", qnode.device.wires)
+
+    execute_kwargs = getattr(qnode, "execute_kwargs", {}).copy()
+    execute_kwargs["cache"] = False
+
+    new_qnode = self.default_qnode_transform(qnode, targs, tkwargs)
+    new_qnode.__class__ = CustomQNode
```

Comment:
this makes me a bit nervous. Is this the most pythonic way to do things?

Reply:
We had a discussion with Edward about that https://github.com/PennyLaneAI/pennylane/pull/4497/files#r1300271100

Reply:
I meant the mechanism for switching classes specifically, rather than the introduction of the `CustomQNode` class. Anyway I lurked on stackoverflow and this seems to be the way, albeit a dangerous one. lgtm